src/server/qgsconfigcache.cpp

```diff
@@ -48,7 +48,7 @@ const QgsProject *QgsConfigCache::project( const QString &path )
       mFileSystemWatcher.addPath( path );
     }
   }
-
+  QgsProject::setInstance( mProjectCache[ path ] );
```

Comment:
It's too bad, we ne longer had QgsProject singleton on server side :(.

Is there no other way out?

Reply:
Well, actually we've never had it, but still there is a single project instance which is the active/current at a given time (per-request), all we need is a way to switch this active instance on each request.

I don't see a way out because of the project caching mechanism in the server: if we want to keep it (and I think we do) we need a way to set the current project.

All the rest of QGIS core assumes that there is a single(ton) project active and we cannot change that.

Btw, I think it's safe to merge this PR because it fixes one serious bug (and maybe more) , we can look after a better solution after.

But I totally agree with you: I also don't like seeing this `setInstance` in public API.

Reply:
OK let's do this so.

But we should keep that in mind for the future :).

Reply:
Can we have a `QgsApplication::instance()->activeProject()` method? I think that one will even survive QGIS 4...

Reply:
@m-kuhn can you elaborate?

Reply:
https://github.com/qgis/QGIS/pull/7248

Reply:
Sounds good, I'll wait for you to merge it.

Reply:
Especially after the latest comment, I'm waiting for broader feedback before merging it.